elastic/kibana
# [Console] Add Console parser to kbn/monaco

## Summary

Related meta [issue](https://github.com/elastic/kibana/issues/176926) 

This PR adds the parser used by Ace in Console to the Console language definition in `kbn/monaco`. 
Changes introduced by this PR: 
- Copy the code for `'sense_editor/mode/worker_parser'` from the file `src/plugins/console/public/application/models/legacy_core_editor/mode/worker/worker.js` into the `kbn/monaco` package
- Move the code for the webworker from the `xjson` folder in `kbn/monaco` to a shared folder `ace_migration`
- Register the parser worker for the Console language in `kbn/monaco`

### How to test
#### Test the parser in Console
1. Add `console.dev.enableMonaco: true` to kibana.dev.yml
2. Open Dev Tools Console and try to type a valid request, check that there are no red markers in the editor
3. Type an invalid request and check that there are red markers in the editor

#### Test that `xjson` language parser still works
1. Navigate to Ingest pipelines and click the "create from csv" button
2. Load a valid csv file, for example this [one](https://github.com/kgeller/ecs-mapper/blob/master/example/mapping.csv) 
3. In the editor that now display a valid json, try changing the value and check that red markers appear for invalid json 

### Screenshots
#### Invalid request (red markers in the editor)
<img width="786" alt="Screenshot 2024-02-19 at 18 06 13" src="https://github.com/elastic/kibana/assets/6585477/bac1bdfd-c402-45f1-9b9b-a9cc29ccb123">


#### Valid request
<img width="795" alt="Screenshot 2024-02-19 at 18 06 23" src="https://github.com/elastic/kibana/assets/6585477/c06b1163-1077-43c6-bddc-1d86d0116266">


### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
